File: app/src/main/java/io/github/summerdez/asmrplayer/ui/activity/DlsiteLoginActivity.java
package io.github.summerdez.asmrplayer.ui.activity;

import io.github.summerdez.asmrplayer.R;
import io.github.summerdez.asmrplayer.data.*;
import io.github.summerdez.asmrplayer.data.remote.*;
import io.github.summerdez.asmrplayer.data.download.*;
import io.github.summerdez.asmrplayer.data.files.*;
import io.github.summerdez.asmrplayer.domain.*;
import io.github.summerdez.asmrplayer.domain.model.*;
import io.github.summerdez.asmrplayer.playback.*;
import io.github.summerdez.asmrplayer.presentation.*;
import io.github.summerdez.asmrplayer.ui.*;
import io.github.summerdez.asmrplayer.ui.activity.*;
import io.github.summerdez.asmrplayer.ui.components.*;
import io.github.summerdez.asmrplayer.ui.screens.*;
import io.github.summerdez.asmrplayer.ui.theme.*;
import io.github.summerdez.asmrplayer.ui.util.*;
import io.github.summerdez.asmrplayer.di.*;
import android.app.Activity;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.activity.OnBackPressedCallback;
import androidx.appcompat.app.AppCompatActivity;

public class DlsiteLoginActivity extends AppCompatActivity {
    private WebView webView;
    private View progressView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getSupportActionBar() != null) {
            getSupportActionBar().hide();
        }
        AppUi.refreshTheme(this);
        AppUi.applySystemBars(this);
        setContentView(buildUi());
        configureWebView();
        registerBackHandler();
        webView.loadUrl(DlsiteClient.LOGIN_URL);
    }

    @Override
    protected void onDestroy() {
        if (webView != null) {
            webView.destroy();
            webView = null;
        }
        super.onDestroy();
    }

    private void registerBackHandler() {
        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                if (webView != null && webView.canGoBack()) {
                    webView.goBack();
                    return;
                }
                finishWithResult();
            }
        });
    }

    private View buildUi() {
        FrameLayout root = new FrameLayout(this);
        root.setBackgroundColor(AppUi.BG);

        LinearLayout content = new LinearLayout(this);
        content.setOrientation(LinearLayout.VERTICAL);
        content.setBackgroundColor(AppUi.BG);
        root.addView(content, new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT));

        LinearLayout header = AppUi.horizontalRow(this);
        header.setGravity(Gravity.CENTER_VERTICAL);
        header.setPadding(dp(16), dp(10), dp(12), dp(8));
        header.setBackgroundColor(AppUi.BG);

        TextView close = AppUi.compactTitleText(this, "关闭");
        close.setTextColor(AppUi.ACCENT);
        close.setTextSize(AppUi.TEXT_BODY);
        close.setGravity(Gravity.CENTER_VERTICAL);
        close.setOnClickListener(view -> finishWithResult());
        header.addView(close, new LinearLayout.LayoutParams(dp(64), dp(44)));

        TextView title = AppUi.compactTitleText(this, "DLsite 登录");
        title.setTextSize(AppUi.TEXT_HEADLINE);
        title.setGravity(Gravity.CENTER);
        title.setTypeface(title.getTypeface(), android.graphics.Typeface.BOLD);
        header.addView(title, new LinearLayout.LayoutParams(
                0,
                LinearLayout.LayoutParams.WRAP_CONTENT,
                1f));

        TextView done = AppUi.compactTitleText(this, "完成");
        done.setTextColor(AppUi.ACCENT);
        done.setTextSize(AppUi.TEXT_BODY);
        done.setGravity(Gravity.CENTER);
        done.setOnClickListener(view -> finishWithResult());
        header.addView(done, new LinearLayout.LayoutParams(dp(64), dp(44)));
        content.addView(header, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                dp(64)));

        progressView = new View(this);
        progressView.setBackgroundColor(AppUi.ACCENT);
        content.addView(progressView, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                Math.max(1, dp(2))));

        webView = new WebView(this);
        content.addView(webView, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                0,
                1f));
        return root;
    }

    private void configureWebView() {
        CookieManager cookieManager = CookieManager.getInstance();
        cookieManager.setAcceptCookie(true);
        cookieManager.setAcceptThirdPartyCookies(webView, true);

        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);
        settings.setLoadWithOverviewMode(true);
        settings.setUseWideViewPort(true);
        settings.setSafeBrowsingEnabled(true);
        settings.setAllowFileAccess(false);
        settings.setAllowContentAccess(false);
        settings.setMixedContentMode(WebSettings.MIXED_CONTENT_NEVER_ALLOW);

        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageStarted(WebView view, String url, android.graphics.Bitmap favicon) {
                progressView.setVisibility(View.VISIBLE);
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                CookieManager.getInstance().flush();
                progressView.setVisibility(View.GONE);
            }
        });
    }

    private void finishWithResult() {
        CookieManager.getInstance().flush();
        setResult(Activity.RESULT_OK);
        finish();
    }

    private int dp(int value) {
        return AppUi.dp(this, value);
    }
}

File: app/src/main/java/io/github/summerdez/asmrplayer/ui/activity/DlsiteLoginActivity.kt
package io.github.summerdez.asmrplayer.ui.activity

import android.app.Activity
import android.graphics.Bitmap
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.webkit.CookieManager
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import io.github.summerdez.asmrplayer.data.remote.DlsiteClient
import io.github.summerdez.asmrplayer.ui.theme.AppUi
import kotlin.math.max

open class DlsiteLoginActivity : AppCompatActivity() {
    private var webView: WebView? = null
    private var progressView: View? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar?.hide()
        AppUi.refreshTheme(this)
        AppUi.applySystemBars(this)
        setContentView(buildUi())
        configureWebView()
        registerBackHandler()
        webView?.loadUrl(DlsiteClient.LOGIN_URL)
    }

    override fun onDestroy() {
        webView?.destroy()
        webView = null
        super.onDestroy()
    }

    private fun registerBackHandler() {
        onBackPressedDispatcher.addCallback(
            this,
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    if (webView?.canGoBack() == true) {
                        webView?.goBack()
                        return
                    }
                    finishWithResult()
                }
            },
        )
    }

    private fun buildUi(): View {
        val root = FrameLayout(this)
        root.setBackgroundColor(AppUi.BG)

        val content = LinearLayout(this)
        content.orientation = LinearLayout.VERTICAL
        content.setBackgroundColor(AppUi.BG)
        root.addView(
            content,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT,
            ),
        )

        val header = AppUi.horizontalRow(this)
        header.gravity = Gravity.CENTER_VERTICAL
        header.setPadding(dp(16), dp(10), dp(12), dp(8))
        header.setBackgroundColor(AppUi.BG)

        val close = AppUi.compactTitleText(this, "关闭")
        close.setTextColor(AppUi.ACCENT)
        close.textSize = AppUi.TEXT_BODY.toFloat()
        close.gravity = Gravity.CENTER_VERTICAL
        close.setOnClickListener { finishWithResult() }
        header.addView(close, LinearLayout.LayoutParams(dp(64), dp(44)))

        val title = AppUi.compactTitleText(this, "DLsite 登录")
        title.textSize = AppUi.TEXT_HEADLINE.toFloat()
        title.gravity = Gravity.CENTER
        title.setTypeface(title.typeface, android.graphics.Typeface.BOLD)
        header.addView(
            title,
            LinearLayout.LayoutParams(
                0,
                LinearLayout.LayoutParams.WRAP_CONTENT,
                1f,
            ),
        )

        val done = AppUi.compactTitleText(this, "完成")
        done.setTextColor(AppUi.ACCENT)
        done.textSize = AppUi.TEXT_BODY.toFloat()
        done.gravity = Gravity.CENTER
        done.setOnClickListener { finishWithResult() }
        header.addView(done, LinearLayout.LayoutParams(dp(64), dp(44)))
        content.addView(
            header,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                dp(64),
            ),
        )

        progressView = View(this)
        progressView?.setBackgroundColor(AppUi.ACCENT)
        content.addView(
            progressView,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                max(1, dp(2)),
            ),
        )

        webView = WebView(this)
        content.addView(
            webView,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                0,
                1f,
            ),
        )
        return root
    }

    private fun configureWebView() {
        val cookieManager = CookieManager.getInstance()
        cookieManager.setAcceptCookie(true)
        cookieManager.setAcceptThirdPartyCookies(webView, true)

        val settings = webView?.settings ?: return
        settings.javaScriptEnabled = true
        settings.domStorageEnabled = true
        settings.databaseEnabled = true
        settings.loadWithOverviewMode = true
        settings.useWideViewPort = true
        settings.safeBrowsingEnabled = true
        settings.allowFileAccess = false
        settings.allowContentAccess = false
        settings.mixedContentMode = WebSettings.MIXED_CONTENT_NEVER_ALLOW

        webView?.webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                progressView?.visibility = View.VISIBLE
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                CookieManager.getInstance().flush()
                progressView?.visibility = View.GONE
            }
        }
    }

    private fun finishWithResult() {
        CookieManager.getInstance().flush()
        setResult(Activity.RESULT_OK)
        finish()
    }

    private fun dp(value: Int): Int {
        return AppUi.dp(this, value)
    }
}
